core/src/main/scala/kafka/controller/ZookeeperClient.scala

```diff
@@ -74,50 +80,64 @@ class ZookeeperClient(connectString: String, sessionTimeoutMs: Int, connectionTi
     }
     val countDownLatch = new CountDownLatch(requests.size)
     val responseQueue = new ArrayBlockingQueue[AsyncResponse](requests.size)
-    requests.foreach {
+    requests.foreach(request => send(request, countDownLatch, responseQueue))
+    countDownLatch.await()
+    responseQueue.asScala.toSeq
+  }
+
+  private def send(request: AsyncRequest, countDownLatch: CountDownLatch, responseQueue: ArrayBlockingQueue[AsyncResponse]): Unit = {
+    import scala.collection.JavaConverters._
+    inFlightRequests.acquire()
```

Comment:
Normally we use try..finally.. block to release the semaphore permit.  otherwise, we may not be releasing the semaphore permits on unexpected errors.